### PR TITLE
feat(query): implement arithmetic functions for `Date` and `Timestamp`.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

**Almost 3000 LOC are unit tests. Feel free to review the codes.**

- [x] Remove `IntervalType`.

Implementation:

- [x] [add|substract]_[years|quarters|months|days|hours|minutes|seconds]
- [x] date_add/date_substract
- [x] arithmetic operator (+/-)

Unit tests:

- [x] [add|substract]_[years|quarters|months|days|hours|minutes|seconds]
- [x] date_add/date_substract
- [x] arithmetic operator (+/-)

Part of #8048

### CODE REVIEW

**Comment by RinChanNOWWW** [comment]
2/3 of the LOC are unit tests.

File: src/query/expression/src/date_helper.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    ))
+}
+
+// Get the last day of the year month, could be 28(non leap Feb), 29(leap year Feb), 30 or 31
+fn last_day_of_year_month(year: i32, month: u32) -> u32 {
+    let is_leap_year = NaiveDate::from_ymd_opt(year, 2, 29).is_some();
+    if std::intrinsics::unlikely(month == 2 && is_leap_year) {
+        return 29;
+    }
+    let last_day_lookup = [0u32, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
```
Comment (sundy-li) [inline src/query/expression/src/date_helper.rs:100]: it could be static variable.

**Comment by Xuanwo** [comment]
> Almost 3000 LOC are unit tests. Feel free to review the codes.

Safe! Very Safe!

File: src/query/functions-v2/tests/it/scalars/parser.rs [hunk 2]
```diff
@@ -21,6 +21,7 @@ use common_ast::parser::token::Token;
 use common_ast::parser::tokenize_sql;
 use common_ast::Backtrace;
 use common_ast::Dialect;
+use common_datavalues::IntervalKind;
```
Comment (b41sh) [inline src/query/functions-v2/tests/it/scalars/parser.rs:24]: can we move `IntervalKind` to `expression`? So that we can remove dependent on `datavalues`.

Comment (RinChanNOWWW) [reply src/query/functions-v2/tests/it/scalars/parser.rs:24]: We can do this. But we need make `ast::parser` use `expression`, too.

Comment (sundy-li) [reply src/query/functions-v2/tests/it/scalars/parser.rs]: IntervalKind should be defined in AST crate. 

And expression crate depends on AST crate.

**Review by sundy-li — APPROVED** [review]

**Review by b41sh — APPROVED** [review]

**Comment by RinChanNOWWW** [comment]
@mergify update

**Comment by RinChanNOWWW** [comment]
@mergify update

**Pull request merged**